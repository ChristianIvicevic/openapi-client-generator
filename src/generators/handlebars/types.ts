import type { OpenAPIV3 } from 'openapi-types';
import type { OperationInfo } from 'parser/types';

export type SchemasCompilationContext = {
  readonly schemaObjects: ReadonlyArray<{
    readonly name: string;
    readonly schemaObject:
      | OpenAPIV3.ReferenceObject
      | OpenAPIV3.ArraySchemaObject
      | OpenAPIV3.NonArraySchemaObject;
  }>;
};

export type RequestsCompilationContext = {
  readonly referencedSchemas: readonly string[];
  readonly operations: readonly OperationInfo[];
};
