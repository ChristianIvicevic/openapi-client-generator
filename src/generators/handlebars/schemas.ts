import { pascalCase } from 'change-case';
import type { OpenAPIV3 } from 'openapi-types';
import { assertIsDefined } from 'utils/assert';
import { isReferenceObject } from 'utils/type-guards';

export const generateSchema = (
  schemaObject:
    | OpenAPIV3.ReferenceObject
    | OpenAPIV3.NonArraySchemaObject
    | OpenAPIV3.ArraySchemaObject,
): string => {
  if (isReferenceObject(schemaObject)) {
    return generateRefSchema(schemaObject.$ref);
  }

  const nullable = schemaObject.nullable ? ' | null' : '';
  return `${generateScalarSchema(schemaObject)}${nullable}`;
};

const generateScalarSchema = (
  schemaObject: OpenAPIV3.ArraySchemaObject | OpenAPIV3.NonArraySchemaObject,
) => {
  switch (schemaObject.type) {
    case 'array':
      return generateArraySchema(schemaObject);

    case 'boolean':
      return 'boolean';

    case 'integer':
    case 'number':
      return 'number';

    /* eslint-disable @typescript-eslint/ban-ts-comment */
    // @ts-ignore
    case 'null':
    // @ts-ignore
    // eslint-disable-next-line no-fallthrough
    case "'null'":
      return 'null';
    /* eslint-enable @typescript-eslint/ban-ts-comment */

    case 'object':
      return generateObjectSchema(schemaObject);

    case 'string':
      return generateStringSchema(schemaObject);

    default:
      return (
        generateCombinedSchema(schemaObject) ??
        generateDictionarySchema(schemaObject) ??
        'unknown'
      );
  }
};

const generateRefSchema = ($ref: OpenAPIV3.ReferenceObject['$ref']) => {
  if ($ref.startsWith('#/components/schemas')) {
    return pascalCase($ref.replace('#/components/schemas/', ''));
  }

  throw Error(
    `The reference '${$ref}' does not match the pattern '#/components/schemas/*'`,
  );
};

const generateArraySchema = ({ items }: OpenAPIV3.ArraySchemaObject) =>
  `readonly (${generateSchema(items)})[]`;

const generateStringSchema = ({
  enum: stringLiterals,
}: OpenAPIV3.NonArraySchemaObject) => {
  if (stringLiterals !== undefined) {
    return stringLiterals.map(literal => `'${String(literal)}'`).join(' | ');
  }

  return 'string';
};

const generateCombinedSchema = (
  schemaObject: OpenAPIV3.NonArraySchemaObject,
) => {
  if (schemaObject.allOf) {
    return generateAllOfSchema(schemaObject);
  }

  if (schemaObject.oneOf) {
    return generateOneOfSchema(schemaObject);
  }

  return undefined;
};

const generateAllOfSchema = ({ allOf }: OpenAPIV3.NonArraySchemaObject) => {
  assertIsDefined(allOf);

  return allOf.map(generateSchema).join(' & ');
};

const generateOneOfSchema = ({ oneOf }: OpenAPIV3.NonArraySchemaObject) => {
  assertIsDefined(oneOf);

  return oneOf.map(generateSchema).join(' | ');
};

const generateObjectSchema = (schemaObject: OpenAPIV3.NonArraySchemaObject) => {
  const { required, properties } = schemaObject;

  if (properties === undefined) {
    return (
      generateCombinedSchema(schemaObject) ??
      generateDictionarySchema(schemaObject) ??
      'unknown'
    );
  }

  const isRequired = (property: string) => (required ?? []).includes(property);

  const objectProperties = Object.entries(properties).map(
    ([name, prop]) =>
      `readonly ${name}${isRequired(name) ? '' : '?'}: ${generateSchema(prop)}`,
  );

  return objectProperties.length === 0
    ? 'unknown'
    : `{\n${objectProperties.join(';')}\n}`;
};

const generateDictionarySchema = ({
  additionalProperties,
}: OpenAPIV3.NonArraySchemaObject) => {
  // TODO: additionalProperties === true is not handled yet
  if (
    additionalProperties !== undefined &&
    typeof additionalProperties !== 'boolean'
  ) {
    const recordType = generateSchema(additionalProperties);
    return `Record<string, ${recordType}>`;
  }

  return undefined;
};
