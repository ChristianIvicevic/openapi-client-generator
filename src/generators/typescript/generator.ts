import { camelCase, pascalCase } from 'change-case';
import { createLeadingTrivia } from 'generators/typescript/factories/leading-trivia';
import { createRequestFunction } from 'generators/typescript/factories/request-function';
import type { TypescriptCompilationContext } from 'generators/typescript/types';
import type { OpenAPIV3 } from 'openapi-types';
import { OperationMethod } from 'parser/types';
import { partial, partition } from 'ramda';
import ts, { factory } from 'typescript';
import { assertIsDefined } from 'utils/assert';
import { format } from 'utils/format';
import { compact } from 'utils/fp';
import { getLogger } from 'utils/logging';
import { isReferenceObject } from 'utils/type-guards';

export const compileUsingTypescript = (document: OpenAPIV3.Document) => {
  const logger = getLogger();
  const referencedSchemas: string[] = [];

  logger.verbose('Compiling schemas...');
  const compiledSchemaNodes = Object.entries(document.components?.schemas ?? {})
    .sort(([a], [b]) => a.localeCompare(b))
    .flatMap(([schema, schemaObject]) => {
      logger.debug(`Compiling schema '${schema}'`);
      return compileSchema(
        { document, referencedSchemas },
        schema,
        schemaObject,
      );
    });

  logger.verbose('Compiling operations...');
  // This block is a workaround to prefetch operation ids in order to sort the
  // generated methods alphanumerically rather than in their order inside the
  // schema file. This is intentional to produce more deterministic output
  // during compilation.
  const compiledOperationNodes = compact(
    Object.entries(document.paths ?? {})
      .flatMap(([path, pathItemObject]) => {
        assertIsDefined(pathItemObject);

        const { parameters } = pathItemObject;

        return compact(
          (['get', 'post', 'put', 'delete'] as OperationMethod[]).map(method =>
            pathItemObject[method]
              ? {
                  context: { document, path, referencedSchemas },
                  method,
                  pathParameters: parameters,
                  operationObject: pathItemObject[method],
                  operationId:
                    pathItemObject[method]?.operationId ?? '<NO OPERATION ID>',
                }
              : undefined,
          ),
        );
      })
      .sort((a, b) => a.operationId?.localeCompare(b.operationId))
      .flatMap(({ context, method, operationObject, pathParameters }) =>
        compileOperation(context, method, pathParameters, operationObject),
      ),
  );

  if (compiledSchemaNodes.length !== 0) {
    ts.addSyntheticLeadingComment(
      compiledSchemaNodes[0],
      ts.SyntaxKind.MultiLineCommentTrivia,
      ' eslint-disable ',
      true,
    );

    ts.addSyntheticLeadingComment(
      compiledSchemaNodes[0],
      ts.SyntaxKind.MultiLineCommentTrivia,
      ' THIS FILE HAS BEEN GENERATED AUTOMATICALLY - DO NOT EDIT IT MANUALLY ',
      true,
    );
  }

  const schemasContent = ts
    .createPrinter({ newLine: ts.NewLineKind.LineFeed })
    .printList(
      ts.ListFormat.MultiLine,
      factory.createNodeArray(compiledSchemaNodes),
      ts.createSourceFile(
        'schemas.ts',
        '',
        ts.ScriptTarget.Latest,
        false,
        ts.ScriptKind.TS,
      ),
    );

  const requestsContent = ts
    .createPrinter({ newLine: ts.NewLineKind.LineFeed })
    .printList(
      ts.ListFormat.MultiLine,
      factory.createNodeArray([
        ...createLeadingTrivia({
          referencedSchemas: [
            ...new Set(referencedSchemas.sort((a, b) => a.localeCompare(b))),
          ],
        }),
        ...compiledOperationNodes,
      ]),
      ts.createSourceFile(
        'requests.ts',
        '',
        ts.ScriptTarget.Latest,
        false,
        ts.ScriptKind.TS,
      ),
    );

  return {
    schemas: format(schemasContent),
    requests: format(requestsContent),
  };
};

const compileSchema = (
  context: TypescriptCompilationContext,
  schemaName: string,
  schemaObject:
    | OpenAPIV3.ReferenceObject
    | OpenAPIV3.ArraySchemaObject
    | OpenAPIV3.NonArraySchemaObject,
) => {
  const description = isReferenceObject(schemaObject)
    ? // TODO: Dereference description of deeply referenced schema object.
      ''
    : schemaObject.description ?? '';
  return compact([
    description.length !== 0 && factory.createJSDocComment(description, []),
    factory.createTypeAliasDeclaration(
      undefined,
      [factory.createModifier(ts.SyntaxKind.ExportKeyword)],
      factory.createIdentifier(pascalCase(schemaName)),
      undefined,
      resolveSchema(context, schemaObject),
    ),
  ]);
};

const compileOperation = (
  context: TypescriptCompilationContext,
  method: OperationMethod,
  pathParameters: readonly (
    | OpenAPIV3.ReferenceObject
    | OpenAPIV3.ParameterObject
  )[] = [],
  operationObject?: OpenAPIV3.OperationObject,
) => {
  const logger = getLogger();
  const { path } = context;

  assertIsDefined(path);

  if (operationObject === undefined) {
    logger.debug(`Skipping empty '${method}' operation object`);
    return undefined;
  }

  const {
    operationId,
    summary,
    responses,
    requestBody,
    parameters,
  } = operationObject;

  if (operationId === undefined) {
    throw Error(`'${method}' operation object has no operation id`);
  }

  logger.debug(`Compiling operation '${operationId}'`);

  const responseTypes = compileResponses(context, responses);
  // TODO: Handle required flag for the request body.
  const requestBodyTypes = compileRequestBody(context, requestBody);
  const {
    pathParameterPropertySignatures,
    queryParameterPropertySignatures,
  } = compileAndPartitionParameters(context, [
    ...pathParameters,
    ...(parameters ?? []),
  ]);

  return createRequestFunction({
    operationId: camelCase(operationId),
    method,
    path,
    summary,
    responseTypes,
    requestBodyTypes,
    pathParameterPropertySignatures,
    queryParameterPropertySignatures,
  });
};

const compileAndPartitionParameters = (
  context: TypescriptCompilationContext,
  parameters: readonly (
    | OpenAPIV3.ReferenceObject
    | OpenAPIV3.ParameterObject
  )[],
) => {
  const { document } = context;
  const [
    pathParameters,
    otherParameters,
  ] = partition<OpenAPIV3.ParameterObject>(
    parameter => parameter.in === 'path',
    parameters.map(parameter => {
      if (!isReferenceObject(parameter)) {
        return parameter;
      }

      const referencedParameter =
        document.components?.parameters?.[
          parameter.$ref.replace('#/components/parameters/', '')
        ];

      if (referencedParameter === undefined) {
        throw Error(`Cannot dereference parameter '${parameter.$ref}'`);
      }

      if (isReferenceObject(referencedParameter)) {
        throw Error(
          `Dereferenced parameter '${parameter.$ref}' is a deep reference to '${referencedParameter.$ref}', which is not supported`,
        );
      }

      return referencedParameter;
    }),
  );

  // TODO: Handle header and cookie parameters.
  const pathParameterPropertySignatures = compileParameterSignatures(
    context,
    pathParameters,
  );
  const queryParameterPropertySignatures = compileParameterSignatures(
    context,
    otherParameters.filter(parameter => parameter.in === 'query'),
  );

  return {
    pathParameterPropertySignatures,
    queryParameterPropertySignatures,
  };
};

const compileParameterSignatures = (
  context: TypescriptCompilationContext,
  parameterObjects: readonly OpenAPIV3.ParameterObject[],
) =>
  parameterObjects.map(parameter =>
    factory.createPropertySignature(
      [factory.createModifier(ts.SyntaxKind.ReadonlyKeyword)],
      factory.createIdentifier(parameter.name),
      parameter.required
        ? undefined
        : factory.createToken(ts.SyntaxKind.QuestionToken),
      parameter.schema === undefined
        ? factory.createKeywordTypeNode(ts.SyntaxKind.NeverKeyword)
        : resolveSchema(context, parameter.schema),
    ),
  );

const compileResponses = (
  context: TypescriptCompilationContext,
  responses?: OpenAPIV3.ResponsesObject,
) => {
  if (responses === undefined) {
    return [factory.createKeywordTypeNode(ts.SyntaxKind.VoidKeyword)];
  }

  const isResponseOk = ([statusCode]: [string, unknown]) =>
    statusCode.toString().startsWith('2') ||
    // We have to consider 'default' responses as well since they can be of any
    // kind, be it OK or ERROR
    statusCode.toString().startsWith('default');

  return Object.entries(responses)
    .filter(isResponseOk)
    .map(([, response]) => response)
    .flatMap(response => compileResponseOrRequestBody(context, response));
};

const compileRequestBody = (
  context: TypescriptCompilationContext,
  requestBody?: OpenAPIV3.ReferenceObject | OpenAPIV3.RequestBodyObject,
) => {
  if (requestBody === undefined) {
    return undefined;
  }

  return compileResponseOrRequestBody(context, requestBody);
};

const compileResponseOrRequestBody = (
  context: TypescriptCompilationContext,
  component:
    | OpenAPIV3.ReferenceObject
    | OpenAPIV3.ResponseObject
    | OpenAPIV3.RequestBodyObject,
) => {
  if (isReferenceObject(component)) {
    // TODO: Implement dereferencing via context.
    throw Error('NYI: Resolving #/components/(responses|requestBodies)/*');
  }

  if (component.content === undefined) {
    return [factory.createKeywordTypeNode(ts.SyntaxKind.VoidKeyword)];
  }

  return Object.entries(component.content).map(([mediaType, mediaTypeObject]) =>
    compileMediaType(context, mediaType, mediaTypeObject),
  );
};

const compileMediaType = (
  context: TypescriptCompilationContext,
  mediaType: string,
  mediaTypeObject: OpenAPIV3.MediaTypeObject,
): ts.TypeNode => {
  const logger = getLogger();

  const supportedMediaTypes = [
    '*/*',
    'application/json',
    'application/octet-stream',
    'application/x-yaml',
  ];

  const { schema } = mediaTypeObject;

  if (
    !supportedMediaTypes.some(supportedMediaType =>
      mediaType.startsWith(supportedMediaType),
    ) ||
    // TODO: This check doesn't seem correct...
    schema === undefined
  ) {
    logger.warn(`Ignoring unsupported content type '${mediaType}'`);
    return factory.createKeywordTypeNode(ts.SyntaxKind.UnknownKeyword);
  }

  return resolveSchema(context, schema);
};

const resolveRefSchema = (
  { referencedSchemas }: TypescriptCompilationContext,
  $ref: OpenAPIV3.ReferenceObject['$ref'],
) => {
  if ($ref.startsWith('#/components/schemas')) {
    const schema = pascalCase($ref.replace('#/components/schemas/', ''));
    referencedSchemas.push(schema);
    return factory.createTypeReferenceNode(factory.createIdentifier(schema));
  }
  throw Error(
    `The reference '${$ref}' does not match the pattern '#/components/schemas/*'`,
  );
};

const resolveSchema = (
  context: TypescriptCompilationContext,
  schemaObject:
    | OpenAPIV3.ReferenceObject
    | OpenAPIV3.NonArraySchemaObject
    | OpenAPIV3.ArraySchemaObject,
): ts.TypeNode => {
  if (isReferenceObject(schemaObject)) {
    return resolveRefSchema(context, schemaObject.$ref);
  }

  return schemaObject.nullable
    ? factory.createUnionTypeNode([
        resolveInlineSchema(context, schemaObject),
        factory.createLiteralTypeNode(factory.createNull()),
      ])
    : resolveInlineSchema(context, schemaObject);
};

const resolveInlineSchema = (
  context: TypescriptCompilationContext,
  schemaObject: OpenAPIV3.NonArraySchemaObject | OpenAPIV3.ArraySchemaObject,
) => {
  switch (schemaObject.type) {
    case 'array':
      return factory.createTypeOperatorNode(
        ts.SyntaxKind.ReadonlyKeyword,
        factory.createArrayTypeNode(resolveSchema(context, schemaObject.items)),
      );
    case 'boolean':
      return factory.createKeywordTypeNode(ts.SyntaxKind.BooleanKeyword);
    case 'integer':
    case 'number':
      return factory.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword);
    // This is a bandaid fix until the openapi-types packages adds null to the
    // types which is available with OpenAPI v3.1
    /* eslint-disable @typescript-eslint/ban-ts-comment */
    // @ts-ignore
    case 'null':
    // @ts-ignore
    // eslint-disable-next-line no-fallthrough
    case "'null'":
      return factory.createLiteralTypeNode(factory.createNull());
    /* eslint-enable @typescript-eslint/ban-ts-comment */
    case 'object':
      return resolveObjectSchema(context, schemaObject);
    case 'string':
      return schemaObject.enum === undefined
        ? factory.createKeywordTypeNode(ts.SyntaxKind.StringKeyword)
        : factory.createUnionTypeNode(
            schemaObject.enum.map(enumValue =>
              factory.createLiteralTypeNode(
                factory.createStringLiteral(String(enumValue)),
              ),
            ),
          );
    default:
      return resolveFallbackSchema(context, schemaObject);
  }
};

const resolveFallbackSchema = (
  context: TypescriptCompilationContext,
  schemaObject: OpenAPIV3.NonArraySchemaObject,
) =>
  tryResolveCombinedSchema(context, schemaObject) ??
  tryResolveDictionarySchema(context, schemaObject) ??
  factory.createKeywordTypeNode(ts.SyntaxKind.UnknownKeyword);

const resolveObjectSchema = (
  context: TypescriptCompilationContext,
  schemaObject: OpenAPIV3.NonArraySchemaObject,
) => {
  const { required, properties } = schemaObject;

  if (properties === undefined) {
    return resolveFallbackSchema(context, schemaObject);
  }

  const isRequired = (property: string) => (required ?? []).includes(property);

  const objectProperties = Object.entries(properties).map(([name, prop]) =>
    factory.createPropertySignature(
      [factory.createModifier(ts.SyntaxKind.ReadonlyKeyword)],
      factory.createIdentifier(name),
      isRequired(name)
        ? undefined
        : factory.createToken(ts.SyntaxKind.QuestionToken),
      resolveSchema(context, prop),
    ),
  );

  return objectProperties.length === 0
    ? factory.createKeywordTypeNode(ts.SyntaxKind.UnknownKeyword)
    : factory.createTypeLiteralNode(objectProperties);
};

const tryResolveCombinedSchema = (
  context: TypescriptCompilationContext,
  schemaObject: OpenAPIV3.NonArraySchemaObject,
) => {
  if (schemaObject.allOf) {
    return factory.createIntersectionTypeNode(
      schemaObject.allOf.map(partial(resolveSchema, [context])),
    );
  }

  if (schemaObject.oneOf) {
    return factory.createUnionTypeNode(
      schemaObject.oneOf.map(partial(resolveSchema, [context])),
    );
  }

  return undefined;
};

const tryResolveDictionarySchema = (
  context: TypescriptCompilationContext,
  schemaObject: OpenAPIV3.NonArraySchemaObject,
) => {
  const { additionalProperties } = schemaObject;

  // TODO: additionalProperties === true is not handled yet for excess properties.
  if (
    additionalProperties !== undefined &&
    typeof additionalProperties !== 'boolean'
  ) {
    return factory.createTypeReferenceNode(factory.createIdentifier('Record'), [
      factory.createKeywordTypeNode(ts.SyntaxKind.StringKeyword),
      resolveSchema(context, additionalProperties),
    ]);
  }

  return undefined;
};
