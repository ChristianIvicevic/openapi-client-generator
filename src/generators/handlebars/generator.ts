import { pascalCase } from 'change-case';
import { generateSchema } from 'generators/handlebars/schemas';
import type {
  RequestsCompilationContext,
  SchemasCompilationContext,
} from 'generators/handlebars/types';
import * as Handlebars from 'handlebars';
import type {
  DocumentInfo,
  OperationInfo,
  OperationMethod,
  ParameterInfo,
  TypeInfo,
} from 'parser/types';
import { format } from 'prettier';
import { getLogger } from 'utils/logging';
import wordWrap from 'word-wrap';

const wrapJsdoc = (text: string) =>
  wordWrap(text, {
    width: 80 - 3,
    newline: '\n * ',
    indent: '',
  });

// BEGIN Handlebars helper functions

const formatParameters = (text: string) => text.replace(/\{/g, '${parameters.');

const getRequestBodyArgument = (
  method: OperationMethod,
  typeInfo?: readonly TypeInfo[],
) => {
  if (method !== 'post' && method !== 'put') {
    return '';
  }

  return typeInfo === undefined ? 'undefined,' : 'requestBody,';
};

const hasQueryParameters = (parameterInfo: ParameterInfo[]) =>
  parameterInfo.some(parameter => parameter.in === 'query');

const hasRequestBody = (typeInfo?: readonly TypeInfo[]) =>
  typeInfo !== undefined;

const isQueryParameter = (parameterInfo: ParameterInfo) =>
  parameterInfo.in === 'query';

const resolveType = ({ isArray, type }: TypeInfo) =>
  isArray ? `readonly ${type}[]` : type;

const resolveTypes = (typeInfo?: readonly TypeInfo[]) => {
  const logger = getLogger();

  // TODO: Figure out whether this is even possible.
  if (typeInfo === undefined) {
    logger.warn(`A type was unexpectedly resolved to 'unknown'`);
    return 'unknown';
  }

  const compiledType = typeInfo.map(resolveType).join(' | ');
  return `(${compiledType})`;
};

Handlebars.registerHelper('format-parameters', formatParameters);
Handlebars.registerHelper('get-request-body-arg', getRequestBodyArgument);
Handlebars.registerHelper('generate-schema', generateSchema);
Handlebars.registerHelper('has-query-parameters', hasQueryParameters);
Handlebars.registerHelper('has-request-body', hasRequestBody);
Handlebars.registerHelper('is-query-parameter', isQueryParameter);
Handlebars.registerHelper('resolve-type', resolveType);
Handlebars.registerHelper('resolve-types', resolveTypes);
Handlebars.registerHelper('wrap-jsdoc', wrapJsdoc);

// END Handlebars helper functions

const prettier = (content: string) =>
  format(content, {
    ...{
      arrowParens: 'avoid',
      singleQuote: true,
      tabWidth: 2,
      trailingComma: 'all',
    },
    parser: 'typescript',
  });

export const compileUsingHandlebars = async ({
  pathItemObjects,
  schemaObjects,
}: DocumentInfo) => {
  // The import MUST have the explicit suffix otherwise the ts path plugin
  // won't be able to resolve it properly for runtime
  await import('generators/handlebars/templates/templates.precompiled.js');

  const schemasTemplate = Handlebars.templates[
    'schemas.hbs'
  ] as HandlebarsTemplateDelegate<SchemasCompilationContext>;

  const requestsTemplate = Handlebars.templates[
    'requests.hbs'
  ] as HandlebarsTemplateDelegate<RequestsCompilationContext>;

  const operations = Object.values(pathItemObjects)
    .map(
      pathItemObject =>
        Object.values(pathItemObject).filter(Boolean) as OperationInfo[],
    )
    .flat()
    .sort((a, b) => a.operationId.localeCompare(b.operationId));

  const referencedTypes = operations.flatMap(
    ({ parameters, requestBodyType, responseType }) => [
      ...parameters.map(({ typeInfo: { type } }) => type),
      ...(requestBodyType?.map(({ type }) => type) ?? []),
      ...responseType.map(({ type }) => type),
    ],
  );

  const referencedSchemas = [
    ...new Set(
      referencedTypes.filter(
        type =>
          ![
            'boolean',
            'number',
            'object',
            'string',
            'unknown',
            'void',
          ].includes(type),
      ),
    ),
  ].sort((a, b) => a.localeCompare(b));

  const parsedSchemaObjects = Object.entries(schemaObjects)
    .map(([name, schemaObject]) => ({ name: pascalCase(name), schemaObject }))
    .sort((a, b) => a.name.localeCompare(b.name));

  const compiledSchemasOutput = schemasTemplate({
    schemaObjects: parsedSchemaObjects,
  });

  const compiledRequestsOutput = requestsTemplate({
    operations,
    referencedSchemas,
  });

  return {
    schemas: prettier(compiledSchemasOutput),
    requests: prettier(compiledRequestsOutput),
  };
};
