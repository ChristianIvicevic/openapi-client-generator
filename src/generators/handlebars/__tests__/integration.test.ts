import { readFileSync } from 'fs';
import { compileUsingHandlebars } from 'generators/handlebars/generator';
import { parseYaml } from 'parser/parser';
import { compose } from 'ramda';

// TESTEE function
const compile = compose(compileUsingHandlebars, parseYaml);

describe('Handlebars Generator Integration', () => {
  it('compiles the link example schema', async () => {
    // GIVEN the link example OpenAPI schema
    const document = readFileSync('test/link-example.yaml').toString();

    // WHEN compiling with the handlebars generator
    const { requests, schemas } = await compile(document);

    // THEN the output matches the snapshot
    expect(requests).toMatchSnapshot();
    expect(schemas).toMatchSnapshot();
  });

  it('compiles the petstore schema', async () => {
    // GIVEN the petstore OpenAPI schema
    const document = readFileSync('test/petstore.yaml').toString();

    // WHEN compiling with the handlebars generator
    const { requests, schemas } = await compile(document);

    // THEN the output matches the snapshot
    expect(requests).toMatchSnapshot();
    expect(schemas).toMatchSnapshot();
  });

  it('compiles the expanded petstore schema', async () => {
    // GIVEN the expanded petstore OpenAPI schema
    const document = readFileSync('test/petstore-expanded.yaml').toString();

    // WHEN compiling with the handlebars generator
    const { requests, schemas } = await compile(document);

    // THEN the output matches the snapshot
    expect(requests).toMatchSnapshot();
    expect(schemas).toMatchSnapshot();
  });

  it('compiles the uspto schema', async () => {
    // GIVEN the uspto OpenAPI schema
    const document = readFileSync('test/uspto.yaml').toString();

    // WHEN compiling with the handlebars generator
    const { requests, schemas } = await compile(document);

    // THEN the output matches the snapshot
    expect(requests).toMatchSnapshot();
    expect(schemas).toMatchSnapshot();
  });
});
