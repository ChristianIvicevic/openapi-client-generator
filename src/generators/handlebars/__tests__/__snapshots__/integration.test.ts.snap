// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handlebars Generator Integration compiles the expanded petstore schema 1`] = `
"/* eslint-disable */
/* THIS FILE HAS BEEN GENERATED AUTOMATICALLY - DO NOT EDIT IT MANUALLY */

import axios, { AxiosRequestConfig } from 'axios';
import type { Error, NewPet, Pet } from './schemas';

/**
 * @param requestBody The payload to send with this request
 * @param config Overrides of the axios configuration for this request
 */
export const addPet = async (
  requestBody: NewPet,
  config?: AxiosRequestConfig,
) => axios.post<Pet | Error>(\`/pets\`, requestBody, config);

/**
 * @param parameters Parameters associated with this request
 * @param config Overrides of the axios configuration for this request
 */
export const deletePet = async (
  parameters: {
    /**
     * ID of pet to delete
     */
    id: number;
  },
  config?: AxiosRequestConfig,
) => axios.delete<void | Error>(\`/pets/\${parameters.id}\`, config);

/**
 * @param parameters Parameters associated with this request
 * @param config Overrides of the axios configuration for this request
 */
export const findPetById = async (
  parameters: {
    /**
     * ID of pet to fetch
     */
    id: number;
  },
  config?: AxiosRequestConfig,
) => axios.get<Pet | Error>(\`/pets/\${parameters.id}\`, config);

/**
 * @param parameters Parameters associated with this request
 * @param config Overrides of the axios configuration for this request
 */
export const findPets = async (
  parameters: {
    /**
     * tags to filter by
     */
    tags: readonly string[];
    /**
     * maximum number of results to return
     */
    limit: number;
  },
  config?: AxiosRequestConfig,
) =>
  axios.get<readonly Pet[] | Error>(\`/pets\`, {
    ...config,
    params: {
      tags: parameters.tags,
      limit: parameters.limit,
    },
  });
"
`;

exports[`Handlebars Generator Integration compiles the link example schema 1`] = `
"/* eslint-disable */
/* THIS FILE HAS BEEN GENERATED AUTOMATICALLY - DO NOT EDIT IT MANUALLY */

import axios, { AxiosRequestConfig } from 'axios';
import type { Pullrequest, Repository, User } from './schemas';

/**
 * @param parameters Parameters associated with this request
 * @param config Overrides of the axios configuration for this request
 */
export const getPullRequestsById = async (
  parameters: {
    /**
     *
     */
    username: string;
    /**
     *
     */
    slug: string;
    /**
     *
     */
    pid: string;
  },
  config?: AxiosRequestConfig,
) =>
  axios.get<Pullrequest>(
    \`/2.0/repositories/\${parameters.username}/\${parameters.slug}/pullrequests/\${parameters.pid}\`,
    config,
  );

/**
 * @param parameters Parameters associated with this request
 * @param config Overrides of the axios configuration for this request
 */
export const getPullRequestsByRepository = async (
  parameters: {
    /**
     *
     */
    username: string;
    /**
     *
     */
    slug: string;
    /**
     *
     */
    state: string;
  },
  config?: AxiosRequestConfig,
) =>
  axios.get<readonly Pullrequest[]>(
    \`/2.0/repositories/\${parameters.username}/\${parameters.slug}/pullrequests\`,
    {
      ...config,
      params: {
        state: parameters.state,
      },
    },
  );

/**
 * @param parameters Parameters associated with this request
 * @param config Overrides of the axios configuration for this request
 */
export const getRepositoriesByOwner = async (
  parameters: {
    /**
     *
     */
    username: string;
  },
  config?: AxiosRequestConfig,
) =>
  axios.get<readonly Repository[]>(
    \`/2.0/repositories/\${parameters.username}\`,
    config,
  );

/**
 * @param parameters Parameters associated with this request
 * @param config Overrides of the axios configuration for this request
 */
export const getRepository = async (
  parameters: {
    /**
     *
     */
    username: string;
    /**
     *
     */
    slug: string;
  },
  config?: AxiosRequestConfig,
) =>
  axios.get<Repository>(
    \`/2.0/repositories/\${parameters.username}/\${parameters.slug}\`,
    config,
  );

/**
 * @param parameters Parameters associated with this request
 * @param config Overrides of the axios configuration for this request
 */
export const getUserByName = async (
  parameters: {
    /**
     *
     */
    username: string;
  },
  config?: AxiosRequestConfig,
) => axios.get<User>(\`/2.0/users/\${parameters.username}\`, config);

/**
 * @param parameters Parameters associated with this request
 * @param config Overrides of the axios configuration for this request
 */
export const mergePullRequest = async (
  parameters: {
    /**
     *
     */
    username: string;
    /**
     *
     */
    slug: string;
    /**
     *
     */
    pid: string;
  },
  config?: AxiosRequestConfig,
) =>
  axios.post<void>(
    \`/2.0/repositories/\${parameters.username}/\${parameters.slug}/pullrequests/\${parameters.pid}/merge\`,
    undefined,
    config,
  );
"
`;

exports[`Handlebars Generator Integration compiles the petstore schema 1`] = `
"/* eslint-disable */
/* THIS FILE HAS BEEN GENERATED AUTOMATICALLY - DO NOT EDIT IT MANUALLY */

import axios, { AxiosRequestConfig } from 'axios';
import type { Error, Pet, Pets } from './schemas';

/**
 * Create a pet
 * @param config Overrides of the axios configuration for this request
 */
export const createPets = async (config?: AxiosRequestConfig) =>
  axios.post<void | Error>(\`/pets\`, undefined, config);

/**
 * List all pets
 * @param parameters Parameters associated with this request
 * @param config Overrides of the axios configuration for this request
 */
export const listPets = async (
  parameters: {
    /**
     * How many items to return at one time (max 100)
     */
    limit: number;
  },
  config?: AxiosRequestConfig,
) =>
  axios.get<Pets | Error>(\`/pets\`, {
    ...config,
    params: {
      limit: parameters.limit,
    },
  });

/**
 * Info for a specific pet
 * @param parameters Parameters associated with this request
 * @param config Overrides of the axios configuration for this request
 */
export const showPetById = async (
  parameters: {
    /**
     * The id of the pet to retrieve
     */
    petId: string;
  },
  config?: AxiosRequestConfig,
) => axios.get<Pet | Error>(\`/pets/\${parameters.petId}\`, config);
"
`;

exports[`Handlebars Generator Integration compiles the uspto schema 1`] = `
"/* eslint-disable */
/* THIS FILE HAS BEEN GENERATED AUTOMATICALLY - DO NOT EDIT IT MANUALLY */

import axios, { AxiosRequestConfig } from 'axios';
import type { DataSetList } from './schemas';

/**
 * List available data sets
 * @param config Overrides of the axios configuration for this request
 */
export const listDataSets = async (config?: AxiosRequestConfig) =>
  axios.get<DataSetList>(\`/\`, config);

/**
 * Provides the general information about the API and the list of fields that
 * can be used to query the dataset.
 * @param parameters Parameters associated with this request
 * @param config Overrides of the axios configuration for this request
 */
export const listSearchableFields = async (
  parameters: {
    /**
     * Name of the dataset.
     */
    dataset: string;
    /**
     * Version of the dataset.
     */
    version: string;
  },
  config?: AxiosRequestConfig,
) =>
  axios.get<string>(
    \`/\${parameters.dataset}/\${parameters.version}/fields\`,
    config,
  );

/**
 * Provides search capability for the data set with the given search criteria.
 * @param parameters Parameters associated with this request
 * @param requestBody The payload to send with this request
 * @param config Overrides of the axios configuration for this request
 */
export const performSearch = async (
  parameters: {
    /**
     * Version of the dataset.
     */
    version: string;
    /**
     * Name of the dataset. In this case, the default value is oa_citations
     */
    dataset: string;
  },
  requestBody: unknown,
  config?: AxiosRequestConfig,
) =>
  axios.post<readonly object[]>(
    \`/\${parameters.dataset}/\${parameters.version}/records\`,
    requestBody,
    config,
  );
"
`;
